include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -48,6 +50,27 @@ namespace gko {
 namespace solver {
 
 
+/**
+ * Traits class providing information on the type and location of workspace
+ * vectors inside a solver.
+ */
+template <typename Solver>
+struct solver_workspace_traits {
```

Comment:
The ids are already always available for Lower/UpperTrs, so I think we really should make the functions constexpr.

Reply:
I see. I guess you mean the current Multigrid design, right? In any case, going from non-constexpr to constexpr shouldn't cause issues later on, so we can leave the interface as it is.
I would still suggest making it constexpr for the Trs, since the ids are already available.

Reply:
making these constexpr actually leads to other problems, because the solver_base implementation tries to call them and complains if they are not implemented in the header, and I would like to avoid moving everything there.